### DISCUSSION TITLE
IPOPT and MATLAB MA57 Dynamic Library

### DISCUSSION THREAD

**Discussion by anilvrao** [coin-or/Ipopt#722] [Q&A]
Dear IPOPT Users Group,

I am trying to compile a new Mex File for IPOPT.  My immediate goal is to compile a Mex file for the Silicon (M) Version of MATLAB, but I would eventually like to compile new Mex files for Intel Mac and Linux.  

In any case, here is where I am with my Mex file compilation and my question.  First, if I use the Harwell source code for MA57 I am able to successfully compile a Mex file.  The problem is that I cannot distribute this Mex file due to the Harwell licensing rules.  I would like to compile a Mex file that will load the run time library "libmwma57.dylib" that is part of MATLAB and is located in the directory $MATLAB/bin/maca64.   When I use the flag "--enable-matlab-ma57" I am able to successfully compile a Mex file, but upon execution I get an error that the library "libhsl.dylib" cannot be found.  I know that MA57 has dependencies, and perhaps libhsl.dylib contains these dependencies, but I do not know how to include these dependencies to create a Mex file independent of the Harwell source code.  

Any help is appreciated.  

Anil Rao

**Comment by svigerske** [comment]
I don't think we have an option `--enable-matlab-ma57` in current Ipopt. The compiler processor reacts on whether `FUNNY_MA57_FINT` is defined and it changes the integer type to be used with MA57. I don't know whether that is still useful.

You could try setting option [hsllib](https://coin-or.github.io/Ipopt/OPTIONS.html#OPT_hsllib) to point to `libmwma57.dylib`.
But if you have a very old Ipopt, then there is no such option and you need to use other means to make it pick up `libmwma57.dylib` when it is looking for `libhsl.dylib`.

However, when building an Ipopt for use with Matlab, there is always the risk of a mixup of Lapack versions, since Matlab uses 64bit integers and Ipopt uses 32bit integers (by default). This is also the case for Matlab's MA57, which is the reason for the `FUNNY_MA57_FINT`.

Please also take a look at https://github.com/ebertolazzi/mexIPOPT and maybe join https://github.com/ebertolazzi/mexIPOPT/issues/24.

**Comment by anilvrao** [comment]
What values does FUNNY_MA57_FINT take?  Is it a string or a number?  Where do I include it in the compiler options once I define its value?

Does the MATLAB MA57 library contain all of the required HSL code?  In other words, if I make a symbolic link from lingual.dylib to libmwma57.dylib, does the MATLAB dynamic library contain the everything that is needed?

Thanks!

— Anil

——
Anil V. Rao, PhD
E-mail: ***@***.***
Tel: 352-672-1529
Web:  http://anilvrao.com

**Reply by svigerske** [reply]
> What values does FUNNY_MA57_FINT take?  Is it a string or a number?  Where do I include it in the compiler options once I define its value?

It just checks whether it is defined. So, `ADD_CXXFLAGS=-DFUNNY_MA57_FINT` should do it.

> Does the MATLAB MA57 library contain all of the required HSL code?  

Probably.

> In other words, if I make a symbolic link from lingual.dylib to libmwma57.dylib, does the MATLAB dynamic library contain the everything that is needed?

Try and see.

**Reply by anilvrao** [reply]
Would I compile with the configure flag “—without-hsl” or with the flag “—with-hsl”?

**Reply by svigerske** [reply]
I think you need `--without-hsl`, since there is only a library (no header file) and the lib has only MA57.

**Reply by anilvrao** [reply]
Where do you put the ADD_CXXFLAGS line?

**Reply by anilvrao** [reply]
I think I now understand that I need to compile the ThirdParty-HSL code into the Mex file without ma57 and then find a way to link to the MATLAB library libmwma57.dylib. I just do not know how to make it so that the Mex file (well, IPOPT) knows to load the libmwma57.dylib library at run time. I am confused as to how to make that connection. Not sure how to properly set flags.

**Reply by svigerske** [reply]
`ADD_CXXFLAGS`  you can set as environment variable or pass as argument to `configure`.

If `libmwma57.dylib` already includes MA57, then you won't have to build another library with HSL codes via ThirdParty-HSL. Also, the latter are not redistributable.
If you don't provide a HSL library when Ipopt is build, but select MA57 as linear solver, Ipopt will try to load a library at runtime that contains MA57 (https://coin-or.github.io/Ipopt/INSTALL.html#LINEARSOLVERLOADER). That's where you want `libmwma57.dylib` to be picked  up.

**Reply by anilvrao** [reply]
Do I need to set the environment variable LD_LIBRARY_PATH in my Unix shell in order to find the libmwma57.dylib library?

**Reply by anilvrao** [reply]
I get the following output when I try to compile the IPOPT Mex file.  Do you know the meaning of those lines with the version numbers?  I do not understand what that means given that I compiled IPOPT on my machine.  

mex -largeArrayDims -Isrc  ./src/ipopt.cc ./src/IpoptInterfaceCommon.cc -Isrc_ipopt_osx/coin-or -DOS_MAC -output bin/osx/ipopt_osx  -L/usr/local/lib -lcoinmumps -lcoinhsl -lipopt -lmetis -lGKlib  /opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/libgfortran.a  /opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/libquadmath.a  /opt/homebrew/opt/libomp/lib/libomp.a  -L/opt/homebrew/Cellar/gcc/13.2.0/lib/gcc/13/gcc/aarch64-apple-darwin22/13/ -lgcc LDFLAGS='$LDFLAGS -Wl,-rpath,. -framework Accelerate -ldl' CXXFLAGS='$CXXFLAGS -Wall -O2 -g' 
Building with 'Xcode Clang++'.
/Users/anilvrao/Documents/MATLAB/ipopt_mex-main/mexIPOPT/toolbox/./src/IpoptInterfaceCommon.cc:1890:15: warning: 'vsprintf' is deprecated: This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of sprintf(3), it is highly recommended that you use vsnprintf(3) instead. [-Wdeprecated-declarations]
      nchar = vsprintf(s,pformat,ap);
              ^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/stdio.h:199:1: note: 'vsprintf' has been explicitly marked deprecated here
__deprecated_msg("This function is provided for compatibility reasons only.  Due to security concerns inherent in the design of sprintf(3), it is highly recommended that you use vsnprintf(3) instead.")
^
/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX14.0.sdk/usr/include/sys/cdefs.h:215:48: note: expanded from macro '__deprecated_msg'
        #define __deprecated_msg(_msg) __attribute__((__deprecated__(_msg)))
                                                      ^
1 warning generated.

ld: warning: -undefined error is deprecated
ld: warning: -undefined error is deprecated
ld: warning: object file (/usr/local/lib/libipopt.a[3](IpJournalist.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[5](IpOptionsList.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[6](IpRegOptions.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[7](IpTaggedObject.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[8](IpUtils.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[9](IpLibraryLoader.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[10](IpBlas.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[11](IpCompoundMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[12](IpCompoundSymMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[13](IpCompoundVector.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[14](IpDenseGenMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[15](IpDenseSymMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[16](IpDenseVector.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[17](IpDiagMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[18](IpExpandedMultiVectorMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[19](IpExpansionMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[20](IpIdentityMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[21](IpLapack.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[22](IpLowRankUpdateSymMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[23](IpMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[24](IpMultiVectorMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[25](IpScaledMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[26](IpSumMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[27](IpSumSymMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[28](IpSymScaledMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[29](IpTransposeMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[30](IpVector.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[31](IpZeroMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[32](IpZeroSymMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[33](IpGenTMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[34](IpSymTMatrix.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[35](IpTripletHelper.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[36](IpAdaptiveMuUpdate.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[37](IpAlgBuilder.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[38](IpAlgorithmRegOp.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[39](IpAugRestoSystemSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[40](IpBacktrackingLineSearch.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[41](IpDefaultIterateInitializer.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[42](IpEquilibrationScaling.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[43](IpExactHessianUpdater.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[44](IpFilter.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[45](IpFilterLSAcceptor.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[47](IpGradientScaling.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[48](IpIpoptAlg.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[49](IpIpoptCalculatedQuantities.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[50](IpIpoptData.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[51](IpIteratesVector.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[52](IpLeastSquareMults.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[53](IpLimMemQuasiNewtonUpdater.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[54](IpLoqoMuOracle.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[55](IpLowRankAugSystemSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[56](IpLowRankSSAugSystemSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[57](IpMonotoneMuUpdate.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[58](IpNLPBoundsRemover.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[59](IpNLPScaling.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[60](IpOptErrorConvCheck.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[61](IpOrigIpoptNLP.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[62](IpOrigIterationOutput.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[63](IpPDFullSpaceSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[64](IpPDPerturbationHandler.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[65](IpPDSearchDirCalc.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[66](IpPenaltyLSAcceptor.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[67](IpProbingMuOracle.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[68](IpQualityFunctionMuOracle.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[69](IpRestoConvCheck.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[70](IpRestoFilterConvCheck.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[71](IpRestoIpoptNLP.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[72](IpRestoIterateInitializer.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[73](IpRestoIterationOutput.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[74](IpRestoMinC_1Nrm.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[75](IpRestoPenaltyConvCheck.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[76](IpRestoRestoPhase.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[77](IpStdAugSystemSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[78](IpTimingStatistics.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[79](IpUserScaling.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[80](IpWarmStartIterateInitializer.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[81](IpLinearSolversRegOp.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[82](IpLinearSolvers.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[83](IpSlackBasedTSymScalingMethod.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[84](IpTripletToCSRConverter.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[86](IpTSymLinearSolver.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[87](IpCGPenaltyCq.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[88](IpCGPenaltyData.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[89](IpCGPenaltyLSAcceptor.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[90](IpCGPenaltyRegOp.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[91](IpCGPerturbationHandler.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[92](IpCGSearchDirCalc.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[93](IpPiecewisePenalty.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[94](IpInterfacesRegOp.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[95](IpIpoptApplication.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[96](IpSolveStatistics.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[101](IpTNLPAdapter.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[103](IpMc19TSymScalingMethod.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[104](IpMa27TSolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[105](IpMa57TSolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[106](IpMa77SolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[107](IpMa86SolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[108](IpMa97SolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)
ld: warning: object file (/usr/local/lib/libipopt.a[109](IpPardisoSolverInterface.o)) was built for newer 'macOS' version (14.0) than being linked (11.0)

**Reply by svigerske** [reply]
> Do I need to set the environment variable LD_LIBRARY_PATH in my Unix shell in order to find the libmwma57.dylib library?

On macOS, `LD_LIBRARY_PATH` has no effect. Maybe you mean to set `DYLD_LIBRARY_PATH`. It depends on your system setup whether your dynamic library loader will find the library without the help of  `DYLD_LIBRARY_PATH` or not.

**Reply by svigerske** [reply]
> I get the following output when I try to compile the IPOPT Mex file. Do you know the meaning of those lines with the version numbers? I do not understand what that means given that I compiled IPOPT on my machine.

It says that you try to build a mex file that targets macOS 11, but link in a library (ipopt) that was build to target macOS 14. So it may not be working on macOS 11 (or 12 or 13). If you are on macOS 14, then this is probably not an issue. I guess that its the mex building system that specifes somewhere that macOS 11 should be targeted.

**Comment by anilvrao** [comment]
My understanding (perhaps incorrect) is that the dependencies for ma57 are contained in the publicly available git repository ThirdParty-HSL.git.  I would think these dependencies are required regardless of whether the ma57 source code is compiled into the Mex file or if ma57 is loaded at run time.  Can one not compile this third party code into IPOPT without including ma57?  If so, how does one compile this third party code into the IPOPT Mex file?

Next, I tried creating a symbolic link from libmwma57.dylib to libhsl.dylib, but that did not work.  Thus, I am not sure what is needed in order to have libmwma57.dylib be recognized as the file that contains ma57.

Thanks!

——
Anil V. Rao, PhD
E-mail: ***@***.***
Tel: 352-672-1529
Web:  http://anilvrao.com

**Reply by svigerske** [reply]
There is no code in `ThirdParty-HSL`. This is just a buildsystem for the HSL solvers. So, if you build ThirdParty-HSL without any HSL solvers, you will end up with an empty library.

> Next, I tried creating a symbolic link from libmwma57.dylib to libhsl.dylib, but that did not work. 

Then check why it "didn't work".
Does Ipopt find the library? Does it contain the MA57 function symbols (ma57ad, ma57bd, etc; https://github.com/coin-or/Ipopt/blob/fdd288d7f56e73c65a17c22d6a1b6076379820c8/src/Algorithm/LinearSolvers/IpMa57TSolverInterface.cpp#L314-L318) Does something go wrong when Ipopt calls a function from that library? Etc, etc.

**Comment by anilvrao** [comment]
How do I check to see if libmwma57.dylib contains the ma57 functions you describe?

——
Anil V. Rao, PhD
E-mail: ***@***.***
Tel: 352-672-1529
Web:  http://anilvrao.com

**Reply by svigerske** [reply]
https://stackoverflow.com/questions/4506121/how-to-print-a-list-of-symbols-exported-from-a-dynamic-library

**Reply by anilvrao** [reply]
This is the output of "nm -g libmama57.dylib".  It seems to have MA57 and all of the dependencies in it.  

I created a symbolic link from libmwma57.dylib to libhsl.dylib.  Unfortunately, MATLAB crashes when I run the Mex file after creating that symbolic link.  If I do not create the symbolic link, MATLAB is unable to find libhsl.dylib.  I have compiled IPOPT with the "--wihtout-hsl" option, so the HSL library needs to be loaded externally.  I am not sure why the libmwma57.dylib file does not work with the Mex file.  

Thanks for any further help you can provide.  

Output of "nm -g libmwma57.dylib":

00000000000042f8 T ___NAGf90_ReAlloc
0000000000003890 T ___NAGf90_aAlloc
0000000000003e78 T ___NAGf90_aDeAlloc
000000000002f6c0 S ___NAGf90_allocatable_ql
000000000002f7c0 S ___NAGf90_allocatable_ql2
000000000002fb80 S ___NAGf90_allocatable_tail
000000000002fb90 S ___NAGf90_allocatable_tail2
                 U ___NAGf90_crash
                 U ___NAGf90_end_write_ifile
                 U ___NAGf90_end_write_sequential
                 U ___NAGf90_establish_format
                 U ___NAGf90_expd
00000000000280f0 D ___NAGf90_fast_deallocation_allowed
                 U ___NAGf90_fmt_write_ch
                 U ___NAGf90_fmt_write_d
                 U ___NAGf90_fmt_write_i4
                 U ___NAGf90_freemem
                 U ___NAGf90_getmem
00000000000038b8 T ___NAGf90_laAlloc
0000000000003f3c T ___NAGf90_laDeAlloc
00000000000042f4 T ___NAGf90_line_ReAlloc
0000000000003868 T ___NAGf90_loAlloc
0000000000003db4 T ___NAGf90_loDeAlloc
0000000000004184 T ___NAGf90_loaDeAlloc
                 U ___NAGf90_logd
0000000000003818 T ___NAGf90_lpAlloc
0000000000003b4c T ___NAGf90_lpDeAlloc
00000000000039a8 T ___NAGf90_lzaAlloc
0000000000003958 T ___NAGf90_lzoAlloc
0000000000003908 T ___NAGf90_lzpAlloc
0000000000003840 T ___NAGf90_oAlloc
0000000000003cf0 T ___NAGf90_oDeAlloc
0000000000004014 T ___NAGf90_oaDeAlloc
000000000002fba0 S ___NAGf90_ordinary_ql
000000000002fca0 S ___NAGf90_ordinary_ql2
0000000000030060 S ___NAGf90_ordinary_tail
0000000000030070 S ___NAGf90_ordinary_tail2
00000000000033c0 T ___NAGf90_pAlloc
00000000000039d0 T ___NAGf90_pDeAlloc
                 U ___NAGf90_pagecolour
0000000000030080 S ___NAGf90_pointer_ql2
0000000000030440 S ___NAGf90_pointer_tail2
                 U ___NAGf90_powd
                 U ___NAGf90_rtcrash
                 U ___NAGf90_rterr
                 U ___NAGf90_scalar_ifilex
                 U ___NAGf90_write_ifile
                 U ___NAGf90_write_seq
0000000000003980 T ___NAGf90_zaAlloc
0000000000003930 T ___NAGf90_zoAlloc
00000000000038e0 T ___NAGf90_zpAlloc
                 U ___stack_chk_fail
                 U ___stack_chk_guard
                 U _bzero
                 U _dgemm_
                 U _dgemv_
                 U _dtpsv_
00000000000055dc T _fd15ad_
                 U _getpagesize
                 U _idamax_
00000000000056b0 T _ma57ad_
00000000000072d4 T _ma57bd_
0000000000009778 T _ma57cd_
000000000000cf90 T _ma57dd_
000000000000edf8 T _ma57ed_
000000000000ff04 T _ma57fd_
000000000000efc8 T _ma57gd_
0000000000017158 T _ma57hd_
0000000000005634 T _ma57id_
000000000000f4ac T _ma57jd_
000000000000f9b8 T _ma57kd_
00000000000100e8 T _ma57ld_
00000000000182bc T _ma57lf_
00000000000189fc T _ma57max_
0000000000010620 T _ma57md_
0000000000010904 T _ma57nd_
0000000000010c28 T _ma57od_
0000000000015d68 T _ma57pd_
000000000000a5b0 T _ma57qd_
000000000000ab68 T _ma57rd_
000000000000c464 T _ma57sd_
000000000000c914 T _ma57td_
000000000000b4d8 T _ma57ud_
0000000000016c00 T _ma57vd_
0000000000015e14 T _ma57wd_
0000000000015f94 T _ma57xd_
0000000000016490 T _ma57yd_
00000000000180d8 T _ma57zd_
0000000000018a28 T _mc21ad_
0000000000018a7c T _mc21bd_
0000000000018e24 T _mc34ad_
0000000000019138 T _mc47ad_
000000000001a2b4 T _mc47bd_
0000000000019108 T _mc47id_
000000000001c680 T _mc59ad_
000000000001d478 T _mc59bd_
000000000001dcb8 T _mc59cd_
000000000001df50 T _mc59dd_
000000000001e130 T _mc59ed_
000000000001e3bc T _mc59fd_
000000000001e8e0 T _mc64ad_
000000000001f9cc T _mc64bd_
00000000000204bc T _mc64dd_
00000000000205ac T _mc64ed_
00000000000206e8 T _mc64fd_
000000000001e8b4 T _mc64id_
00000000000214dc T _mc64qd_
0000000000020938 T _mc64rd_
0000000000020bcc T _mc64sd_
0000000000021660 T _mc64ud_
0000000000021b5c T _mc64wd_
0000000000022880 T _mc71ad_
                 U _memcpy
                 U _memset
                 U _memset_pattern16
                 U _metis_nodend_
                 U dyld_stub_binder

**Reply by anilvrao** [reply]
So my current state is that I am able to compile a Mex file without HSL, but I cannot get the Mex file to run. MATLAB either returns that it is unable to find libhsl.dylib, or crashes if I set a symbolic link from libmwma57.dylib to libhsl.dylib.  Makes me wonder if the file libmwma57.dylib in the Silicon version of MATLAB is compatible with the Silicon architecture.

**Reply by svigerske** [reply]
I can reproduce a segfault with Ipopt when using MA57 from `libmwma57.dylib`. Since Matlab used to use 64bit integers, I then enable `FUNNY_MA57_INT` and fix some compilation issues (a8e153f). With this, it seems to run.

I didn't try building mex files or running from Matlab. Others are the experts for this.
